cumulus/client/consensus/aura/src/collators/slot_based/slot_timer.rs

```diff
@@ -0,0 +1,258 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Cumulus. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::LOG_TARGET;
+use codec::Codec;
+use cumulus_primitives_aura::Slot;
+use cumulus_primitives_core::BlockT;
+use sc_client_api::UsageProvider;
+use sc_consensus_aura::SlotDuration;
+use sp_api::ProvideRuntimeApi;
+use sp_application_crypto::AppPublic;
+use sp_consensus_aura::AuraApi;
+use sp_core::Pair;
+use sp_runtime::traits::Member;
+use sp_timestamp::Timestamp;
+use std::{
+	cmp::{max, min},
+	sync::Arc,
+	time::Duration,
+};
+
+/// Lower limits of allowed block production interval.
+/// Defensive mechanism, corresponds to 12 cores at 6 second block time.
+const BLOCK_PRODUCTION_MINIMUM_INTERVAL_MS: Duration = Duration::from_millis(500);
+
+#[derive(Debug)]
+pub(crate) struct SlotInfo {
+	pub timestamp: Timestamp,
+	pub slot: Slot,
+}
+
+/// Manages block-production timings based on chain parameters and assigned cores.
+#[derive(Debug)]
+pub(crate) struct SlotTimer<Block, Client, P> {
+	client: Arc<Client>,
+	drift: Duration,
+	last_reported_core_num: Option<u32>,
+	relay_slot_duration: Duration,
+	_marker: std::marker::PhantomData<(Block, Box<dyn Fn(P) + Send + Sync + 'static>)>,
+}
+
+/// Compute when to try block-authoring next.
+/// The exact time point is determined by the slot duration of relay- and parachain as
+/// well as the last observed core count. If more cores are available, we attempt to author blocks
+/// for them.
+///
+/// Returns a tuple with:
+/// - `Duration`: How long to wait until the next slot.
+/// - `Timestamp`: The timestamp to pass to the inherent
+/// - `Slot`: The AURA slot used for authoring
+fn compute_next_wake_up_time(
+	para_slot_duration: SlotDuration,
+	relay_slot_duration: Duration,
+	core_count: Option<u32>,
+	time_now: Duration,
+	drift: Duration,
+) -> (Duration, Timestamp, Slot) {
+	let para_slots_per_relay_block =
+		(relay_slot_duration.as_millis() / para_slot_duration.as_millis() as u128) as u32;
+	let assigned_core_num = core_count.unwrap_or(1);
+
+	// Trigger at least once per relay block, if we have for example 12 second slot duration,
+	// we should still produce two blocks if we are scheduled on every relay block.
+	let mut block_production_interval = min(para_slot_duration.as_duration(), relay_slot_duration);
+
+	if assigned_core_num > para_slots_per_relay_block &&
+		para_slot_duration.as_duration() >= relay_slot_duration
+	{
+		block_production_interval =
+			max(relay_slot_duration / assigned_core_num, BLOCK_PRODUCTION_MINIMUM_INTERVAL_MS);
+		tracing::debug!(
+			target: LOG_TARGET,
+			?block_production_interval,
+			"Expected to produce for {assigned_core_num} cores but only have {para_slots_per_relay_block} slots. Attempting to produce multiple blocks per slot."
```

Comment:
```suggestion
			"warning: Expected to produce for {assigned_core_num} cores but only have {para_slots_per_relay_block} slots. Attempting to produce multiple blocks per slot."
```

Reply:
This is an expected scenario, not really a warning. The message is purely for debugging purposes.